contracts/router/tree/TokenTree.sol

```diff
@@ -0,0 +1,578 @@
+    /// as the children of the given node. Given node needs to represent a token from the pool.
+    function _addPool(
+        uint256 nodeIndex,
+        address pool,
+        address poolModule,
+        uint256 tokensAmount
+    ) internal {
+        require(nodeIndex < _nodes.length, "Out of range");
+        Node memory node = _nodes[nodeIndex];
+        if (poolModule == address(0)) poolModule = address(this);
+        (bool wasAdded, uint8 poolIndex) = (false, _poolMap[pool].index);
+        if (poolIndex == 0) {
+            poolIndex = uint8(_pools.length);
+            require(poolIndex <= type(uint8).max, "Too many pools");
+            _pools.push(pool);
+            _poolMap[pool] = Pool({module: poolModule, index: poolIndex});
+            wasAdded = true;
+        } else {
+            // Check if the existing pool could be added to the node. This enforces some sanity checks,
+            // as well the invariant that any path from root to node doesn't contain the same pool twice.
+            _checkPoolAddition(nodeIndex, node.depth, poolIndex);
+        }
+        // Remember that the pool is attached to the node
+        _attachedPools[nodeIndex] |= 1 << poolIndex;
+        address[] memory tokens = _getPoolTokens(poolModule, pool, tokensAmount);
+        bool nodeFound = false;
+        unchecked {
+            uint8 childDepth = node.depth + 1;
+            uint256 rootPathParent = _rootPath[nodeIndex];
+            for (uint256 i = 0; i < tokensAmount; ++i) {
+                address token = tokens[i];
+                // Save token indexes if this is a new pool
+                if (wasAdded) {
+                    tokenIndexes[pool][token] = uint8(i);
+                }
+                // Add new nodes to the tree
+                if (token == node.token) {
+                    nodeFound = true;
+                    continue;
+                }
+                _addNode(token, childDepth, poolIndex, rootPathParent);
+            }
+        }
+        require(nodeFound, "Node token not found in the pool");
+    }
+
+    /// @dev Adds a new node to the tree and saves its path to the root.
+    function _addNode(
+        address token,
+        uint8 depth,
+        uint8 poolIndex,
+        uint256 rootPathParent
+    ) internal {
+        // Index of the newly inserted child node
+        uint256 nodeIndex = _nodes.length;
+        require(nodeIndex <= type(uint8).max, "Too many nodes");
+        // Don't add the bridge token (root) twice. This may happen if we add a new pool containing the bridge token
+        // to a few existing nodes. E.g. we have old nUSD/USDC/USDT pool, and we add a new nUSD/USDC pool. In this case
+        // we attach nUSD/USDC pool to root, and then attach old nUSD/USDC/USDT pool to the newly added USDC node
+        // to enable nUSD -> USDC -> USDT swaps via new + old pools.
+        if (_nodes.length > 0 && token == _nodes[0].token) {
```

Comment:
test for this edge case?

Reply:
Nice catch, the setup was there, but there was no `test_X` function covering that. Fixed in 2e35549b5f2c019016b08050e0981df1bff876a3.